File: C#2/Homeworks/06-Strings-and-Text-Processing/05ParseTags/ParseTags.cs
/*
 You are given a text. Write a program that changes the text in all regions surrounded by the tags  
 <upcase>  and  </upcase>  to upper-case.

Input
•On the only line you will receive a string - the text

Output
•Print the changed string on one line

Constraints
•The tags will not be nested.
•String length will be <= 10000
•Time limit: 0.1s
•Memory limit: 16MB

 * */

using System;
using System.Text;
using System.Text.RegularExpressions;
class ParseTags
{
   static void Main()
   {
       string text = Console.ReadLine();
       //string text = "We are living in a <upcase> yellow submarine </upcase>. We don't have <upcase> anything </upcase> else.";

       string startTag ="<upcase>";
       string endTag ="</upcase>";

       //find first occurences
       int startTagIndex = text.IndexOf(startTag);
       int endTagIndex = text.IndexOf(endTag);
       int nextIndexToSearch = 0;

       while (startTagIndex != -1 && endTagIndex != -1)
       {
           //find text to make uppercase
           int upperTextStart = startTagIndex + startTag.Length + 1;
           int upperTextLength = endTagIndex - upperTextStart - 1;

           string textToUp = text.Substring(upperTextStart, upperTextLength);
           string textToInsert = textToUp.ToUpper();

           //replace the upper text
           text = text.Replace(textToUp, textToInsert);

           //remove tags - lenght +1 for the additional space between the 2 tags
           //text = text.Remove(endTagIndex-1, endTag.Length+1);
           //text = text.Remove(startTagIndex, startTag.Length+1);

           text = text.Remove(endTagIndex, endTag.Length);
           text = text.Remove(startTagIndex, startTag.Length);

           //change indexes
           nextIndexToSearch = 0;
           startTagIndex = text.IndexOf(startTag,nextIndexToSearch);
           endTagIndex = text.IndexOf(endTag,nextIndexToSearch);
           
       }

       Console.WriteLine(text);

       //REGULAR EXPRESSIONS SOLUTIONS

       string text2 = "We are living in a <upcase> yellow submarine </upcase>. We don't have <upcase> anything </upcase> else.";
       string[] sentences = text2.Split('.');
       string res = string.Empty;

       for (int i = 0; i < sentences.Length; i++)
       {
           res = sentences[i].Trim().Replace("<upcase>", string.Empty);
           res = res.Replace("</upcase>", string.Empty);
           //Console.WriteLine(res);
       }
           
        
       //string result = Regex.Replace(text, @"<upcase>(.?*)</upcase>");

       
       /*
       private static string TagsToUpper(string text)
        {
            return Regex.Replace(text, @"<upcase>(.*?)</upcase>", delegate(Match match)
            {
                string current = match.ToString().ToUpper();
                return Regex.Replace(current, @"<[^>]*>", String.Empty);
            });
        }
        */ 

   }
}

File: JS/HW/Strings/parse-tags.js
// You are given a text. Write a function that changes the text in all regions:

// <upcase>text</upcase> to uppercase.
// <lowcase>text</lowcase> to lowercase
// <orgcase>text</orgcase> does not change casing
// Note: Regions can be nested.

function parseTags(args){
    let str = args[0],
        strNoTags = [],
        indexUpOpen = -1,
        indexUpClose = -1,
        indexLowOpen = -1,
        indexLowClose = -1,
        indexOrgOpen = -1,
        indexOrgClose = -1,
        sentences = str.split("."),
        noTags ="";

    console.log(sentences);

     for (var i = 0; i < sentences.length; i++) {

         var start=0, strToCopy ="";

         //no tags
         if(sentences[i].indexOf("<")===-1){
            strNoTags.push(sentences[i]); 
            noTags += sentences[i] + ".\n";
         }
         else
         {
            
            //orgcase
            indexOrgOpen = sentences[i].indexOf("<orgcase>");
            if(indexOrgOpen !== -1){            
                result = sentences[i].trim().replace("<orgcase>","").replace("</orgcase>","");
                
            }

            //upcase
            indexUpOpen = sentences[i].indexOf("<upcase>");
            indexUpClose = sentences[i].indexOf("</upcase>");
            if(indexUpOpen !== -1 && indexUpClose !== -1){
                start = indexUpOpen + 8; 
                strToCopy = sentences[i].substring(start,indexUpClose);              
                result = result.trim().replace("<upcase>","").replace(strToCopy,strToCopy.toUpperCase()).replace("</upcase>","");
                //strNoTags.push(result);
                
            }

            //lowcase
            indexLowOpen = sentences[i].indexOf("<lowcase>");
            indexLowClose = sentences[i].indexOf("</lowcase>");
                if(indexLowOpen !== -1 && indexLowClose){
                start = indexLowOpen + 9;  
                strToCopy = sentences[i].substring(start,indexLowClose);
                result = result.trim().replace("<lowcase>","").replace(strToCopy,strToCopy.toLowerCase()).replace("</lowcase>","");
                
                //strNoTags.push(result);
                }
            noTags += result + ".\n";

        }
            

         /*
         if(sentences[i].indexOf("<")===-1){
            strNoTags.push(sentences[i]); 
         }
         else{ //there is a tag
            
            indexOrgOpen = words[i].indexOf("<orgcase>");
            if(indexOrgOpen !== -1){
             //find the end
             indexOrgClose = words[i].indexOf("</orgcase>");
             console.log(indexOrgOpen + "-" + indexOrgClose);
             start = indexOrgOpen + 9;  
             strToCopy = words[i].substring(start,indexOrgClose);
             //console.log(strToCopy);
             strNoTags.push(strToCopy);         
            }
            indexLowOpen = words[i].indexOf("<lowcase>");
            if(indexLowOpen !== -1){
                 indexLowClose = words[i].indexOf("</lowcase>");
             //console.log(indexLowOpen + "-" + indexLowClose);
             start = indexLowOpen + 9;  
             strToCopy = words[i].substring(start,indexLowClose);
             //console.log(strToCopy.toLowerCase());
             strNoTags.push(strToCopy.toLowerCase());
             }

             indexUpOpen = words[i].indexOf("<upcase>");
             if(indexUpOpen !== -1){
                 indexUpClose = words[i].indexOf("</upcase>");
             //console.log(indexUpOpen + ":" + indexUpClose);
                 if(indexUpClose !== -1) {
                    start = indexUpOpen + 9;  
                    strToCopy = words[i].substring(start,indexUpClose); 
                 }else{
                    strToCopy = words[i].substring(start-1);      
                 }

             //console.log(strToCopy.toUpperCase());
             strNoTags.push(strToCopy.toUpperCase());
             }
             */
        //}
    }
     //console.log(strNoTags.join(".\n"));
     console.log(noTags);
}

//parseTags([ 'We are <orgcase>liViNg</orgcase> in a <upcase>yellow submarine</upcase>. We <orgcase>doN\'t</orgcase> have <lowcase>anything</lowcase> else.' ]);
parseTags([ 'Hello.We are <orgcase>liViNg</orgcase> in a <upcase>yellow submarine</upcase>. We doN\'t have <lowcase>Anything</lowcase> else.' ]);
//result :
//We are liViNg in a YELLOW SUBMARINE. We doN't have anything else.